akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/StashManagement.scala

```diff
@@ -45,34 +62,75 @@ private[akka] trait StashManagement[C, E, S] {
     }
   }
 
-  protected def tryUnstash(
-    behavior: Behavior[InternalProtocol]): Behavior[InternalProtocol] = {
-    if (stashBuffer.nonEmpty) {
-      if (setup.settings.logOnStashing) setup.log.debug("Unstashing message: [{}]", stashBuffer.head.getClass)
-
-      stashBuffer.unstash(setup.context, behavior.asInstanceOf[Behavior[InternalProtocol]], 1, ConstantFun.scalaIdentityFunction)
+  /**
+   * `tryUnstashOne` is called at the end of processing each command (or when persist is completed)
+   */
+  protected def tryUnstashOne(behavior: Behavior[InternalProtocol]): Behavior[InternalProtocol] = {
+    val buffer =
+      if (stashState.isUnstashAllInProgress)
+        stashState.userStashBuffer
+      else
+        stashState.internalStashBuffer
+
+    if (buffer.nonEmpty) {
+      if (setup.settings.logOnStashing) setup.log.debug(
+        "Unstashing message from {} stash: [{}]",
+        if (buffer eq stashState.internalStashBuffer) "internal" else "user", buffer.head)
+
+      if (stashState.isUnstashAllInProgress)
+        stashState.decrementUnstashAllProgress()
```

Comment:
Is this always a pre-check to `stashState.decrementUnstashAllProgress` or only here? I mean is it something that should not be allowed unless `isUnstashAllInProgress` and it should be a guard vs only separate?

Reply:
Looks like it's better inside `decrementUnstashAllProgress`, fixed in 9596cd6